### PR TITLE
add SPLIT_HAND_MATRIX_GRID support

### PR DESCRIPTION
**Pull request by mtei**
## Description

Some keyboards have unused intersections in the key matrix. If a diode is connected to this intersection, it can be used to determine the  handedness. The advantage of this method is that it does not use any additional GPIO pins on the MCU.

It is used as follows.

```c
// keyboards/KEYBOARD_NAME/config.h
//     "this is keyboards/helix/rev2/config.h sample"
#define SPLIT_HAND_MATRIX_GRID D7,B2
```

Also, any unused intersections in the key matrix can be used as DIP switches. I plan to implement it in another PR.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by mtei** [comment]
> A lof of limitations stem from peek_matrix partially re-implementing matrix scan.
> 
> For a less general solution `is_keyboard_left` is weak, so this _could_ be put at the keyboard level (and a lot of these concerns get deferred to when the code is copy pasted to another board).

I think that overriding `is_keyboard_left()` and implementing a partial matrix scan within it is unacceptable, as it creates the possibility of inconsistency with `matrix.c` in the future.

I will also use this approach for DIPSW support.

So, I think peek_matrix() is better to be in `matrix.c`.

**Comment by zvecr** [comment]
> So, I think peek_matrix() is better to be in matrix.c.

personally i think the exact opposite.

I would rather see this implemented as 2 config options, where you specify the 2 pins to use. Then we input/output them and read the pin state.

**Comment by mtei** [comment]
> I would rather see this implemented as 2 config options, where you specify the 2 pins to use. Then we input/output them and read the pin state.

OK, I reimplemented it just the way you wanted it.

**Comment by drashna** [comment]
I'm curious, how does this differ from: 
https://docs.qmk.fm/#/feature_split_keyboard?id=handedness-by-pin

**Comment by mtei** [comment]
For example, let's say you've designed a new keyboard and assigned all the GPIO pins to the matrix, WS2812, encoders, etc. In this case, there is no GPIO pin to assign to SPLIT_HAND_PIN. If there is a KC_NO in the matrix, then SPLIT_HAND_MATRIX_GRID can be used.

**Comment by eucalyn** [comment]
I want this feature!!

File: quantum/split_common/split_util.c [hunk 1]
```diff
@@ -47,11 +50,34 @@ bool waitForUsb(void) {
...
+}
+#endif
+
 __attribute__((weak)) bool is_keyboard_left(void) {
 #if defined(SPLIT_HAND_PIN)
     // Test pin SPLIT_HAND_PIN for High/Low, if low it's right hand
     setPinInput(SPLIT_HAND_PIN);
     return readPin(SPLIT_HAND_PIN);
+#elif defined(SPLIT_HAND_MATRIX_GRID)
+#   ifdef SPLIT_HAND_MATRIX_GRID_LOW_IS_RIGHT
```
Comment (drashna) [inline quantum/split_common/split_util.c:76]: This should be documented too.

Comment (mtei) [reply quantum/split_common/split_util.c:76]: OK. done.

**Comment by drashna** [comment]
I hate to ask, but could this be added to the split keyboard doc too? 
https://github.com/qmk/qmk_firmware/blob/master/docs/feature_split_keyboard.md#setting-handedness

**Comment by mtei** [comment]
I did. I hope my writing is correct.

**Comment by Erovia** [comment]
There is a conflict. Can you rebase and resolve it?

**Comment by mtei** [comment]
Thanks, fixed it.

**Comment by Funkest** [comment]
This improves my keyboard experience at multiple places much more than SPLIT_HAND_PIN.
This function needs no additional implementations different from other detection methods.
I merged this on my branch very easily, but I hope this pull request will be merged strongly.

File: quantum/split_common/split_util.c [hunk 2]
```diff
@@ -47,11 +48,34 @@ bool waitForUsb(void) {
...
+    return peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
+#   else
+    return ! peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
```
Comment (nooges) [inline quantum/split_common/split_util.c:77]:
```suggestion
    return !peek_matrix_intersection(SPLIT_HAND_MATRIX_GRID);
```

**Comment by BlueDrink9** [comment]
Apologies if this is an inappropriate place to raise this, but I find the documentation for this quite confusing and I'd like to help update it.

@mtei would you mind answering some questions?

1. I think this requires rewiring, but the docs don't make this clear. Do you need to connect the diode as if bypassing a switch? I.e. for col2row, connect it from the output (ie unused matrix position column) wire/pin with cathode towards the input (ie unused row) wire/pin? I think a diagram would make this a lot clearer.
2. Does this diode need to be added on both halves? What happens if the unused position is only unused on one half, e.g. a thumb cluster that is on column 5 on the left, but column 1 on the right?

**Comment by mtei** [comment]
For example, suppose you have the following keyboard.
```c    
#define MATRIX_ROW_PINS { D4, C6, D7, E6, B4 }
#define MATRIX_COL_PINS { F4, F5, F6, F7, B1, B3, B2 }

/* COL2ROW, ROW2COL*/
#define DIODE_DIRECTION COL2ROW
```
IMG_4592

You can add a diode to the circuit as below and further configuration as below.

```c
/* Split hand configration */
#define SPLIT_HAND_MATRIX_GRID D4,B2
```
IMG_4593

**Pull request merged**